PR: https://github.com/odoo/odoo/pull/232855

From: 81caaaa3f7212ad5d9d72cd446b65c5fb0a9bef9
From: Moamen(msho)
Date: 2026-03-27 23:39:59

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'purchase.order'

Total Changes: 31

[FIX] purchase_stock: fix empty picking creation by po.

This commit fixes the problem of creating an empty picking when
confirming a PO with negative quantity for the product.

Before this commit:
When creating a PO for a vendor with a negative quantity for a product,
it creates an OUT picking with that same quantity to the vendor as
a return which is correct. However, it also creates an empty IN picking
which is not needed.

After this commit:
Only the OUT picking is created for the order lines with negative
quantity as a return and if there are order lines with positive quantity,
it creates a normal IN picking as expected.

Also as a side improvement, Now all updates to purchase order lines are
reflected correctly to the existing po's pickings.

Enterprise PR: https://github.com/odoo/enterprise/pull/108890

Task-5074899

closes odoo/odoo#232855

Related: odoo/enterprise#108890
Signed-off-by: Steve Van Essche <svs@odoo.com>

================================= pseudo patch: =================================

--- a/addons/project_purchase_stock/models/__init__.py
+++ b/addons/project_purchase_stock/models/__init__.py
@@ -1,4 +1,4 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from . import purchase_order
+from . import stock_move
 from . import stock_rule

--- a/addons/project_purchase_stock/models/purchase_order.py
+++ b/None
@@ -1,16 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models
-
-
-class PurchaseOrder(models.Model):
-    _inherit = 'purchase.order'
-
-    def _prepare_picking(self):
-        res = super()._prepare_picking()
-        if not self.project_id:
-            return res
-        return {
-            **res,
-            'project_id': self.project_id.id,
-        }

--- a/None
+++ b/addons/project_purchase_stock/models/stock_move.py
@@ -0,0 +1,13 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    def _get_new_picking_values(self):
+        return {
+            **super()._get_new_picking_values(),
+            'project_id': self.purchase_line_id.order_id.project_id.id,
+        }
